src/lib/features/instance-stats/getReadOnlyUsers.e2e.test.ts

```diff
@@ -26,22 +141,72 @@ afterAll(async () => {
     await db.destroy();
 });
 
-test('should count user with Viewer root role and no events', async () => {
-    const [user] = await db
-        .rawDatabase('users')
-        .insert({
-            email: 'viewer@example.com',
-            name: 'Viewer User',
-            is_system: false,
-            is_service: false,
-        })
-        .returning('id');
+test('counts viewer without write events or permissions', async () => {
+    const viewer = await createViewer();
+    await addEventForUser(viewer.id, FEATURE_FAVORITED);
 
-    await db.rawDatabase('role_user').insert({
-        role_id: 3,
-        user_id: user.id,
-        project: '*',
-    });
+    await expect(getReadOnlyUsers()).resolves.toEqual(1);
+});
+
+test('counts multiple viewers that satisfy read-only conditions', async () => {
+    await createViewer();
+    await createViewer();
+
+    await expect(getReadOnlyUsers()).resolves.toEqual(2);
+});
+
+test('ignores deleted, system, and service viewers', async () => {
+    await createViewer();
+    await createViewer({ is_system: true });
+    await createViewer({ is_service: true });
+    await createViewer({ deleted_at: new Date() });
+
+    await expect(getReadOnlyUsers()).resolves.toEqual(1);
+});
+
+test('ignores viewers with additional root permissions', async () => {
+    await createViewer();
+    const userWithAdmin = await createViewer();
+    await assignRoleToUser(userWithAdmin.id, adminRootRoleId);
+
+    await expect(getReadOnlyUsers()).resolves.toEqual(1);
+});
+
+test('ignores viewers with project role permissions', async () => {
```

Comment:
Any project role permission? What if the role permission is the same as that of the viewer role?

Reply:
Viewer roles have no permissions.

This is why we add an empty Reader project role in https://github.com/Unleash/unleash/pull/11178, so that you can add a RO user to a private project.